File: hw/gualotuna/u3/hw30/Strategy/src/main/java/model/QuickSort.java
package model;

import controller.SortingStrategy;

/**
 *
 * @author  Brayan Gualotuña  
 */
public class QuickSort implements SortingStrategy {

    @Override
    public int[] sort(int[] array) {
        quickSort(array, 0, array.length - 1);
        return array;
    }

    private void quickSort(int[] array, int low, int high) {
        if (low < high) {
            int pi = partition(array, low, high);

            // Recursively sort elements before and after partition
            quickSort(array, low, pi - 1);
            quickSort(array, pi + 1, high);
        }
    }

    private int partition(int[] array, int low, int high) {
        int pivot = array[high];
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (array[j] <= pivot) {
                i++;

                // Swap array[i] and array[j]
                int temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }

        // Swap array[i + 1] and array[high] (or pivot)
        int temp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = temp;

        return i + 1;
    }

}

File: hw/cañarte/u3/HW30Observer&Strategy/Python/StrategyPattern/ec_edu_sortingapp_controller/quick_sort.py
import sys
import os

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from sorting_strategy import SortingStrategy

class QuickSort(SortingStrategy):
    def sort(self, data):
        self._quicksort(data, 0, len(data) - 1)
        return data

    def _quicksort(self, data, low, high):
        if low < high:
            pi = self._partition(data, low, high)
            self._quicksort(data, low, pi - 1)
            self._quicksort(data, pi + 1, high)

    def _partition(self, data, low, high):
        pivot = data[high]
        i = low - 1
        for j in range(low, high):
            if data[j] <= pivot:
                i += 1
                data[i], data[j] = data[j], data[i]

        data[i + 1], data[high] = data[high], data[i + 1]
        return i + 1
